src/detect-base64-decode.c

```diff
@@ -0,0 +1,710 @@
+#if 0
+    if (det_ctx->base64_decoded_len) {
+        printf("Decoded data:\n");
+        PrintRawDataFp(stdout, det_ctx->base64_decoded,
+            det_ctx->base64_decoded_len);
+    }
+#endif
+
+    return det_ctx->base64_decoded_len > 0;
+}
+
+static int DetectBase64DecodeParse(const char *str, uint32_t *bytes,
+    uint32_t *offset, uint8_t *relative)
+{
+    static const int max = 30;
+    int ov[max];
+    int pcre_rc;
+    const char *bytes_str = NULL;
+    const char *offset_str = NULL;
+    const char *relative_str = NULL;
+    int retval = 0;
+
+    *bytes = 0;
+    *offset = 0;
+    *relative = 0;
+
+    pcre_rc = pcre_exec(decode_pcre, decode_pcre_study, str, strlen(str), 0, 0,
+        ov, max);
+    if (pcre_rc < 3) {
+        goto error;
+    }
+
+    if (pcre_rc >= 3) {
+        if (pcre_get_substring((char *)str, ov, max, 2, &bytes_str) > 0) {
+            if (ByteExtractStringUint32(bytes, 10, 0, bytes_str) <= 0) {
+                SCLogError(SC_ERR_INVALID_RULE_ARGUMENT,
+                    "Bad value for bytes: \"%s\"", bytes_str);
+                goto error;
+            }
+        }
+     }
+
+    if (pcre_rc >= 5) {
+        if (pcre_get_substring((char *)str, ov, max, 4, &offset_str)) {
+            if (ByteExtractStringUint32(offset, 10, 0, offset_str) <= 0) {
+                SCLogError(SC_ERR_INVALID_RULE_ARGUMENT,
+                    "Bad value for offset: \"%s\"", offset_str);
+                goto error;
+            }
+        }
+    }
+
+    if (pcre_rc >= 6) {
+        if (pcre_get_substring((char *)str, ov, max, 5, &relative_str)) {
+            if (strcmp(relative_str, "relative") == 0) {
+                *relative = 1;
+            }
+            else {
+                SCLogError(SC_ERR_INVALID_RULE_ARGUMENT,
+                    "Invalid argument: \"%s\"", relative_str);
+                goto error;
+            }
+        }
+    }
+
+    retval = 1;
+error:
+    if (bytes_str != NULL) {
+        pcre_free_substring(bytes_str);
+    }
+    if (offset_str != NULL) {
+        pcre_free_substring(offset_str);
+    }
+    if (relative_str != NULL) {
+        pcre_free_substring(relative_str);
+    }
+    return retval;
+}
+
+static int DetectBase64DecodeSetup(DetectEngineCtx *de_ctx, Signature *s,
+    char *str)
+{
+    uint32_t bytes = 0;
+    uint32_t offset = 0;
+    uint8_t relative = 0;
+    DetectBase64Decode *data;
+    int sm_list;
+    SigMatch *sm = NULL;
+    SigMatch *pm = NULL;
+
+    if (str != NULL) {
+        if (!DetectBase64DecodeParse(str, &bytes, &offset, &relative)) {
+            goto error;
+        }
+    }
+    data = SCCalloc(1, sizeof(DetectBase64Decode));
+    if (unlikely(data == NULL)) {
+        goto error;
+    }
+    data->bytes = bytes;
+    data->offset = offset;
+    data->relative = relative;
+
+    if (s->list != DETECT_SM_LIST_NOTSET) {
+        sm_list = s->list;
+        if (data->relative) {
+            pm = SigMatchGetLastSMFromLists(s, 4,
```

Comment:
clang catch that?  I can't get that warning myself.  Its obviously correct, just trying to get the warning myself for future use.